crates/circuit/src/dag_circuit.rs

```diff
@@ -6501,6 +6509,184 @@ impl DAGCircuit {
 
         Ok(new_nodes)
     }
+
+    /// Alternative constructor to build an instance of [DAGCircuit] from a `QuantumCircuit`.
+    pub(crate) fn from_circuit(
+        py: Python,
+        qc: QuantumCircuitData,
+        copy_op: bool,
+        qubit_order: Option<Vec<Bound<PyAny>>>,
+        clbit_order: Option<Vec<Bound<PyAny>>>,
+    ) -> PyResult<DAGCircuit> {
+        // Extract necessary attributes
+        let qc_data = qc.data;
+        let num_qubits = qc_data.num_qubits();
+        let num_clbits = qc_data.num_clbits();
+        let num_ops = qc_data.__len__();
+        let num_vars = qc.declared_vars.len() + qc.input_vars.len() + qc.captured_vars.len();
+
+        // Build DAGCircuit with capacity
+        let mut new_dag = DAGCircuit::with_capacity(
+            py,
+            num_qubits,
+            num_clbits,
+            Some(num_vars),
+            Some(num_ops),
+            None,
+        )?;
+
+        // Assign other necessary data
+        new_dag.name = qc.name.map(|ob| ob.unbind());
+
+        // Avoid manually acquiring the GIL.
+        new_dag.global_phase = match qc_data.global_phase() {
+            Param::ParameterExpression(exp) => Param::ParameterExpression(exp.clone_ref(py)),
+            Param::Float(float) => Param::Float(*float),
+            _ => unreachable!("Incorrect parameter assigned for global phase"),
+        };
+
+        if let Some(calibrations) = qc.calibrations {
+            new_dag.calibrations = calibrations;
+        }
+
+        new_dag.metadata = qc.metadata.map(|meta| meta.unbind());
+
+        // Add the qubits depending on order.
+        let mut qubit_map: HashMap<Qubit, Bound<PyAny>> = HashMap::new();
+        if let Some(qubit_ordering) = qubit_order {
+            for qubit in qubit_ordering {
+                if new_dag.qubits.find(&qubit).is_some() {
+                    return Err(DAGCircuitError::new_err(format!(
+                        "duplicate qubits {}",
+                        &qubit
+                    )));
+                }
+                qubit_map.insert(new_dag.add_qubit_unchecked(py, &qubit)?, qubit);
+            }
+        } else {
+            for qubit in qc_data.qubits().bits() {
+                let bound = qubit.clone_ref(py).into_bound(py);
+                qubit_map.insert(new_dag.add_qubit_unchecked(py, qubit.bind(py))?, bound);
+            }
+        }
```

Comment:
I'm pretty sure you could do this as an `iter().map().collect()` instead. Something like:
```suggestion
        let qubit_map: HashMap<Qubit, Bound<PyAny>> = if let Some(qubit_ordering) = qubit_order {
            qubit_ordering.iter().map(|qubit| {
                if new_dag.qubits.find(&qubit).is_some() {
                    Err(DAGCircuitError::new_err(format!(
                        "duplicate qubits {}",
                        &qubit
                    )))
                } else {
                    Ok((new_dag.add_qubit_unchecked(py, &qubit)?, qubit))
                }
            }
        } else {
            qc_data.qubits().bits().map(|qubit| {
                let bound = qubit.clone_ref(py).into_bound(py);
                Ok((new_dag.add_qubit_unchecked(py, qubit.bind(py))?, bound))
            })
        }?;
```

(completely untested I wrote this in the webui). If this works (I'm not sure it will) then using this both `qubit_map` and `clbit_map` would probably be slightly more efficient, but likely not measurably.

Reply:
Oh also the same comment about maybe building a vec here instead of hashmap. When I see `Qubit` as a key my first thought is always "Can we use a vec here instead?"

Reply:
I will try to build a `Vec` instead. The one issue I kept finding is that some circuits would have indices that aren't necessarily in order (Specifically with certain `ControlFlowOp` blocks). I will test it out and update it.